pkg/controllers/clusteroperator/reconcile_test.go

```diff
@@ -0,0 +1,165 @@
+package clusteroperator
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/client-go/config/clientset/versioned/fake"
+	configinformers "github.com/openshift/client-go/config/informers/externalversions"
+	common "github.com/openshift/ibm-roks-toolkit/pkg/controllers"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+func TestReconcile(t *testing.T) {
+	versions := map[string]string{
+		"release":    "alpha",
+		"kubernetes": "beta",
+	}
+
+	testCases := []struct {
+		testCase    string
+		existingCOs []runtime.Object
+	}{
+		{
+			testCase: "when there's no cluster operators",
+		},
+		{
+			testCase: "when a clusterOperator exists with no status",
+			existingCOs: []runtime.Object{
+				&configv1.ClusterOperator{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "openshift-apiserver",
+					},
+				},
+			},
+		},
+		{
+			testCase: "when a clusterOperator exists with outdated versions",
+			existingCOs: []runtime.Object{
+				&configv1.ClusterOperator{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "openshift-apiserver",
+					},
+					Status: configv1.ClusterOperatorStatus{
+						Conditions: nil,
+						Versions: []configv1.OperandVersion{
+							{
+								Name:    "openshift-apiserver",
+								Version: "outdated",
+							},
+							{
+								Name:    "operator",
+								Version: "outdated",
+							},
+						},
+						RelatedObjects: nil,
+						Extension:      runtime.RawExtension{},
+					},
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.testCase, func(t *testing.T) {
+			fakeClient := fake.NewSimpleClientset(tc.existingCOs...)
+			informerFactory := configinformers.NewSharedInformerFactory(fakeClient, common.DefaultResync)
+			stopCh := make(chan struct{})
+			go informerFactory.Start(stopCh)
+			go informerFactory.Config().V1().ClusterOperators().Informer().Run(stopCh)
+
+			for _, obj := range tc.existingCOs {
+				informerFactory.Config().V1().ClusterOperators().Informer().GetStore().Add(obj)
+			}
+
+			r := &ControlPlaneClusterOperatorSyncer{
+				Versions: versions,
+				Client:   fakeClient,
+				Lister:   informerFactory.Config().V1().ClusterOperators().Lister(),
+				Log:      ctrl.Log.WithName("TestReconcile"),
+			}
+
+			request := ctrl.Request{
+				NamespacedName: types.NamespacedName{
+					Name: "openshift-apiserver",
+				},
+			}
+
+			var reconcileError error
+			// we retry here as otherwise there's a race between the fakeclient and the lister
+			// which result in a transient error:
+			// "cannot fetch cluster operator openshift-apiserver: clusteroperator.config.openshift.io "openshift-apiserver" not found"
+			err := wait.PollImmediate(time.Second, 5*time.Second, func() (bool, error) {
```

Comment:
See comment above, change to:
```go
_, err := r.Recconcile(request)
```

Reply:
I think the race is still possible since in `Reconcile` after `ensureClusterOperatorsExist` the call to `r.Lister.Get(req.Name)` might still race. Though with your suggestions is most unlikely.